### PR TITLE
Warn about copies in storage that might overwrite unexpectedly.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/2434

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -363,6 +363,35 @@ void TypeChecker::checkLibraryRequirements(ContractDefinition const& _contract)
 			m_errorReporter.typeError(var->location(), "Library cannot have non-constant state variables");
 }
 
+void TypeChecker::checkDoubleStorageAssignment(Assignment const& _assignment)
+{
+	TupleType const& lhs = *dynamic_cast<TupleType const*>(type(_assignment.leftHandSide()).get());
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: Is it worth adding asserts here?

**Comment by axic** [comment]
Can you add a test which actually corrupts the data?

I think this should do it?
```
contract A {
  uint8 a;
  uint248 b;

  function f() {
    (a, b) = (b, a);
  }
}
```

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Oh sorry, this should either be a different dynamic cast, or an assertion, right.

**Comment by chriseth** [comment]
You mean and end-to-end test and check that data is corrupted (I actually would not call it corrupted, it is just that the order of assignments might be un-intuitive, or rather the programmer might expect the use of a temporary)? Your example would not do - value types are not affected by this behaviour.

**Comment by axic** [comment]
Yes an end-to-end test expecting the corrupted value.

**Review by axic — CHANGES_REQUESTED** [review]
Please add two tests:
- tuple assignments for non-storage
- tuple assignments with pointers

Since the check is disabled for those.

**Review by axic — APPROVED** [review]

**Pull request merged**